[jit] Add options to Operator to enable registration of alias analysis passes

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #18588 [jit] Add Custom graph fusion
* #18846 [jit] Only require python print on certain namespaces
* **#18589 [jit] Add options to Operator to enable registration of alias analysis passes**

Differential Revision: [D14901379](https://our.internmc.facebook.com/intern/diff/D14901379)

bwasti committed:
[jit] Add options to Operator to enable registration of alias analysis passes

bwasti commented on torch/csrc/jit/operator.h:
```diff
@@ -10,6 +10,8 @@
 #include <ATen/ATen.h>
 #include <ATen/core/function_schema.h>
 
+#include <torch/csrc/jit/passes/alias_analysis.h>
```
this might be bad, what do you think @zdevito ?

zdevito commented on torch/csrc/jit/operator.h:
```diff
@@ -79,15 +82,32 @@ struct TORCH_API Operator {
+        op_(std::make_shared<Operation>(std::move(op))),
+        options_(std::move(options)) {}
+
+  struct Options {
```
I would not make this struct part of the Operator class. I want to move away from exposing users directly to the Operator class (and instead have them directly call `.op` with the op implementation and options). So I think Options should be its own class. Putting it in its own header will help separate out the depenendencies (i.e. the options rely on alias_analysis, but the operator header does not rely directly on it).

zdevito commented on torch/csrc/jit/operator.h:
```diff
@@ -79,15 +82,32 @@ struct TORCH_API Operator {
+
+  struct Options {
+    Options() {};
+    using AliasAnalysisFunction = void (AliasDb::*)(Node*);
```
Making AliasAnalysisFunction a pointer rather than an enum doesn't buy us anything here, but it does strongly couple it to the alias_analysis.h header. How about just an enum in a separate header:

```
enum AliasAnalysisKind {
   CREATOR,
   EXTRACTOR, 
}

struct Options {
};
```

We can add more wildly customizable versions by patching options when necessary.

suo replied:
I agree. The `analyzeFoo` functions in alias analysis are not really designed for outside consumption and are probably not clear to users. I think we could have an enum that covers the most common cases:
- Purely functional (no aliases, no mutation on inputs).
- View (output aliases input)
- Free for all (all inputs may be written to, all outputs may alias any input)
- Look at the schema (reads a user-provided schema, allows people to do whatever fancy thing).

bwasti replied:
>reads a user-provided schema
@suo Not sure I follow how this would work

suo replied:
To be clear, you don't have to put all of those options in this PR (just wanted to enumerate some common cases we'll want to add later).

For "Look at the schema"—people defining a custom op can pass a JIT schema if they want; so we can instruct the alias analysis pass to use that schema to populate alias info (similar to how we do with builtin ops).

zdevito reviewed (commented):
Looks pretty good. Small nits.

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

driazati commented on torch/csrc/jit/custom_operator.h:
```diff
@@ -170,7 +170,8 @@ FunctionSchema inferAndCheckSchema(const std::string& schemaOrName) {
     const std::string& schemaOrName,
-    Implementation&& implementation) {
+    Implementation&& implementation,
+    OperatorOptions options = OperatorOptions()) {
```
These should be exposed in `RegisterOperators::op()` as well since that's the suggested way to register ops in the docs/tutorials

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

bwasti committed:
Update on "[jit] Add options to Operator to enable registration of alias analysis passes"

[jit] Add options to Operator to enable registration of alias analysis passes

gh-metadata: pytorch pytorch 18589 gh/bwasti/3/head

suo commented:
Looks like this broke the build?
